File: src/main/java/com/gmail/borlandlp/minigamesdtools/lobby/ServerLobby.java
package com.gmail.borlandlp.minigamesdtools.lobby;

import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools;
import com.gmail.borlandlp.minigamesdtools.arena.ArenaBase;
import com.gmail.borlandlp.minigamesdtools.creator.DataProvider;
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.Hotbar;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

public abstract class ServerLobby {
    private Map<Player, ArenaBase> players = new HashMap<>(); // Player -> Arena[may be null -> if null - player is not arena player]
    private String id;
    private Location spawnPoint;
    private String hotbarID; // may be null

    public void broadcastMessage(String message, String eqw) {
        for (Player player : this.players.keySet()) {
            if(!this.playerPlaysInAnyArena(player)) {
                // send msg
            }
        }
    }

    public boolean transferPlayer(@NotNull Player p, @NotNull ServerLobby l) throws Exception {
        Debug.print(Debug.LEVEL.NOTICE, "transfer Player[name:" + p.getName() + "] from ServerLobby[ID:" + this.getID() + "] to ServerLobby[ID:" + l.getID() + "]");

        if(this.unregisterPlayer(p)) {
            if(l.registerPlayer(p)) {
                return true;
            } else {
                this.registerPlayer(p);
                return false;
            }
        } else {
            return false;
        }
    }

    public void spawn(Player p) {
        p.teleport(this.getSpawnPoint());
    }

    public void setPlayerArena(Player p, ArenaBase a) {
        this.players.put(p, a);
    }

    private boolean applyHotbar(Player p) {
        Hotbar hotbar = null;
        try {
            hotbar = MinigamesDTools.getInstance().getHotbarCreatorHub().createHotbar(this.hotbarID, new DataProvider() {{
                this.set("player", p);
            }});
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }

        if(hotbar == null) {
            Debug.print(Debug.LEVEL.WARNING, "Build of Hotbar[ID:" + this.hotbarID + "] for Player[name:" + p.getName() + "] is null! This is a build error.");
            return false;
        } else {
            MinigamesDTools.getInstance().getHotbarAPI().bindHotbar(hotbar, p);
        }

        return true;
    }

    public void handlePlayerLeaveArena(Player p) throws Exception {
        if(!this.players.containsKey(p)) {
            throw new Exception("Lobby[ID:" + this.getID() + "] doesn't contain a player:" + p.getName());
        }

        this.spawn(p);
        this.applyHotbar(p);
    }

    public boolean registerPlayer(@NotNull Player p) {
        Debug.print(Debug.LEVEL.NOTICE, "register Player[name:" + p.getName() + "] in ServerLobby[ID:" + this.getID() + "]");

        this.players.put(p, null);
        this.spawn(p);

        if(this.hotbarID != null) {
            if(!this.applyHotbar(p)) {
                return false;
            }
        }

        return true;
    }

    public boolean unregisterPlayer(@NotNull Player p) {
        Debug.print(Debug.LEVEL.NOTICE, "unregister Player[name:" + p.getName() + "] in ServerLobby[ID:" + this.getID() + "]");
        if(this.hotbarID != null) {
            try {
                MinigamesDTools.getInstance().getHotbarAPI().unbindHotbar(p);
            } catch (Exception ex) {
                ex.printStackTrace();
                return false;
            }
        }

        this.players.remove(p);

        return true;
    }

    public boolean contains(Player p) {
        return this.players.containsKey(p);
    }

    public boolean playerPlaysInAnyArena(Player p) {
        return this.players.get(p) != null;
    }

    public String getID() {
        return id;
    }

    public void setID(String id) {
        this.id = id;
    }

    public Location getSpawnPoint() {
        return spawnPoint;
    }

    public void setSpawnPoint(Location spawnPoint) {
        this.spawnPoint = spawnPoint;
    }

    public String getHotbarID() {
        return hotbarID;
    }

    public void setHotbarID(String hotbarID) {
        this.hotbarID = hotbarID;
    }

    public abstract void onRegister();
    public abstract void onUnregister();
    public abstract void tickUpdate();
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/lobby/ServerLobby.kt
package com.gmail.borlandlp.minigamesdtools.lobby

import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools
import com.gmail.borlandlp.minigamesdtools.arena.ArenaBase
import com.gmail.borlandlp.minigamesdtools.creator.DataProvider
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.Hotbar
import org.bukkit.Location
import org.bukkit.entity.Player
import java.util.*

abstract class ServerLobby {
    private val players: MutableMap<Player, ArenaBase?> =
        HashMap() // Player -> Arena[may be null -> if null - player is not arena player]
    var id: String? = null
    var spawnPoint: Location? = null
    var hotbarID: String? = null

    fun broadcastMessage(message: String) {
        for (player in players.keys) {
            if (!playerPlaysInAnyArena(player)) {
                player.sendMessage(message)
            }
        }
    }

    @Throws(Exception::class)
    fun transferPlayer(p: Player, lobby: ServerLobby): Boolean {
        Debug.print(
            Debug.LEVEL.NOTICE,
            "transfer Player[name:" + p.name + "] from ServerLobby[ID:" + id + "] to ServerLobby[ID:" + lobby.id + "]"
        )
        return if (unregisterPlayer(p)) {
            if (lobby.registerPlayer(p)) {
                true
            } else {
                registerPlayer(p)
                false
            }
        } else {
            false
        }
    }

    fun spawn(p: Player) {
        p.teleport(spawnPoint)
    }

    fun setPlayerArena(p: Player, a: ArenaBase?) {
        players[p] = a
    }

    /**
     * TODO: Зачем здесь это?
     */
    private fun applyHotbar(player: Player): Boolean {
        var hotbar: Hotbar?
        try {
            hotbar = MinigamesDTools.instance!!.hotbarCreatorHub!!
                .createHotbar(hotbarID!!, object : DataProvider() {
                    init {
                        this["player"] = player
                    }
                })
        } catch (e: Exception) {
            e.printStackTrace()
            return false
        }
        MinigamesDTools.instance!!.hotbarAPI!!.bindHotbar(hotbar, player)
        return true
    }

    @Throws(Exception::class)
    fun handlePlayerLeaveArena(p: Player) {
        if (!players.containsKey(p)) {
            throw Exception("Lobby[ID:" + id + "] doesn't contain a player:" + p.name)
        }
        spawn(p)
        if (hotbarID != null) {// TODO: Refactor - fire event?
            applyHotbar(p)
        }
    }

    fun registerPlayer(p: Player): Boolean {
        Debug.print(
            Debug.LEVEL.NOTICE,
            "register Player[name:" + p.name + "] in ServerLobby[ID:" + id + "]"
        )
        players[p] = null
        spawn(p)
        if (hotbarID != null) {
            if (!applyHotbar(p)) { // TODO: Refactor
                return false
            }
        }
        return true
    }

    fun unregisterPlayer(p: Player): Boolean {
        Debug.print(
            Debug.LEVEL.NOTICE,
            "unregister Player[name:" + p.name + "] in ServerLobby[ID:" + id + "]"
        )
        if (hotbarID != null) {
            try {
                MinigamesDTools.instance!!.hotbarAPI!!.unbindHotbar(p) // TODO: Refactor
            } catch (ex: Exception) {
                ex.printStackTrace()
                return false
            }
        }
        players.remove(p)
        return true
    }

    operator fun contains(p: Player?): Boolean {
        return players.containsKey(p)
    }

    fun playerPlaysInAnyArena(p: Player?): Boolean {
        return players[p] != null
    }

    abstract fun onRegister()
    abstract fun onUnregister()
    abstract fun tickUpdate()
}